File: step_by_step_directions_from_a_binary_tree_node_to_another/StepByStepDirectionsFromABinaryTreeNodeToAnother.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class StepByStepDirectionsFromABinaryTreeNodeToAnother {
    public String getDirections(TreeNode root, int startValue, int destValue) {
        if (root == null || startValue == destValue) {
            return "";
        }

        TreeNode lca = findLCA(root, startValue, destValue);
        if (lca == null) {
            return "";
        }

        String[] path = new String[]{null};
        if (lca.val == startValue) {
            getDirectionsToChild(lca, destValue, new StringBuilder(), path);
            return path[0];
        }
        if (lca.val == destValue) {
            getDirectionsToParent(lca, startValue, new StringBuilder(), path);
            return new StringBuilder(path[0]).reverse().toString();
        }

        String[] path2 = new String[]{null};
        getDirectionsToParent(lca, startValue, new StringBuilder(), path);
        getDirectionsToChild(lca, destValue, new StringBuilder(), path2);
        return path[0] + path2[0];
    }

    private void getDirectionsToChild(TreeNode root, int destValue, StringBuilder sb, String[] path) {
        if (root == null|| path[0] != null) {
            return;
        }
        if (root.val == destValue ) {
            path[0] = sb.toString();
            return;
        }
        
        sb.append("L");
        getDirectionsToChild(root.left, destValue, sb, path);
        sb.deleteCharAt(sb.length() - 1);
        sb.append("R");
        getDirectionsToChild(root.right, destValue, sb, path);
        sb.deleteCharAt(sb.length() - 1);
    }

    private void getDirectionsToParent(TreeNode root, int destValue, StringBuilder sb, String[] path) {
        if (root == null|| path[0] != null) {
            return;
        }
        if (root.val == destValue ) {
            path[0] = sb.toString();
            return;
        }
        
        sb.append("U");
        getDirectionsToParent(root.left, destValue, sb, path);
        getDirectionsToParent(root.right, destValue, sb, path);
        sb.deleteCharAt(sb.length() - 1);
    }

    private TreeNode findLCA(TreeNode root, int startValue, int destValue) {
        if (root == null || root.val == startValue || root.val == destValue) {
            return root;
        }

        TreeNode left = findLCA(root.left, startValue, destValue);
        TreeNode right = findLCA(root.right, startValue, destValue);
        if (left != null && right != null) {
            return root;
        }
        return left == null ? right : left;
    }
}

File: step_by_step_directions_from_a_binary_tree_node_to_another/Step_By_Step_Directions_From_A_Binary_Tree_Node_To_Another.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Step_By_Step_Directions_From_A_Binary_Tree_Node_To_Another {
public:
    string getDirections(TreeNode* root, int startValue, int destValue) {
        if (root == NULL) {
            return "";
        }

        TreeNode* lca = find_lca(root, startValue, destValue);
        if (lca == NULL) {
            return "";
        }
        if (lca->val == startValue) {
            string path = "";
            get_directions_to_child(lca, destValue, path);
            return path;
        }
        if (lca->val == destValue) {
            string path = "";
            get_directions_to_parent(lca, startValue, path);
            return path;
        }

        string path1 = "";
        get_directions_to_parent(lca, startValue, path1);
        string path2 = "";
        get_directions_to_child(lca, destValue, path2);
        return path1 + path2;
    }

private:
    bool get_directions_to_parent(TreeNode* root, int destValue, string& path) {
        if (root == NULL) {
            return false;
        }
        if (root->val == destValue) {
            return true;
        }
        
        path.push_back('U');
        if (get_directions_to_parent(root->left, destValue, path)) {
            return true;
        }
        if (get_directions_to_parent(root->right, destValue, path)) {
            return true;
        }
        path.erase(path.size() - 1);

        return false;
    }

    bool get_directions_to_child(TreeNode* root, int destValue, string& path) {
        if (root == NULL) {
            return false;
        }
        if (root->val == destValue) {
            return true;
        }
        
        path.push_back('L');
        if (get_directions_to_child(root->left, destValue, path)) {
            return true;
        }
        path.erase(path.size() - 1);
        path.push_back('R');
        if (get_directions_to_child(root->right, destValue, path)) {
            return true;
        }
        path.erase(path.size() - 1);
        return false;
    }

    TreeNode* find_lca(TreeNode* root, int startValue, int destValue) {
        if (root == NULL || root->val == startValue || root->val == destValue) {
            return root;
        }

        TreeNode* left = find_lca(root->left, startValue, destValue);
        TreeNode* right = find_lca(root->right, startValue, destValue);
        if (left != NULL && right != NULL) {
            return root;
        }
        return left == NULL ? right : left;
    }
};